### PR TITLE
Loading starter assets in Applab

### PR DESCRIPTION
**Pull request by maddiedierker**
Initial PR for [LP-609](https://codedotorg.atlassian.net/browse/LP-609).
[Technical design](https://docs.google.com/document/d/1KrAAv4klJhLL_siQo-t6F-2zJrCOOub6Z7W2b9IqDto/edit#heading=h.gbynda1ya0gp)

### What it does
- If the `<AssetManager/>` is provided with the `levelName` prop, we will request the starter assets associated with that level from `/level_starter_assets/:level_name`. **This prop is currently always undefined, so this PR should have no impact on users.**
- `/level_starter_assets/:level_name` returns an array of summarized file objects present in S3's `cdo-v3-assets/starter_assets` directory ([view starter_assets dir here](https://s3.console.aws.amazon.com/s3/buckets/cdo-v3-assets/starter_assets/?region=us-east-1&tab=overview)) via the level's `starter_assets` property, which will be structured as follows:
```ruby
starter_assets: {
  # key - friendly file name
  # value - UUID file name stored in S3 (to avoid name collisions)
  "welcome.png" => "134354543.png"
}
```
- `/level_starter_assets/:level_name/:filename` will stream the image data from S3 for the requested file present at `cdo-v3-assets/starter_assets/:filename`. This is used when we display the starter asset in `<AssetThumbnail/>`.

Basically, this PR will allow us to request and display starter assets from S3 in Applab.

### Important note
In order to have starter assets consistent across all environments, they are served from the same bucket/directory (e.g., there isn't a "starter_assets_staging" or "_production" directory). **For this reason, levelbuilders will only be allowed to upload and delete starter assets on levelbuilder while in start mode.** There may be better ways to achieve this same outcome, but I couldn't think of any that weren't much more complicated, so let me know your thoughts!

### Examples
As a levelbuilder, when I visit a level and enter start mode, I can see the starter assets associated with that level:
Screen Shot 2019-08-09 at 9 56 30 AM

As a student, when I visit a level, I can see the starter assets and my own uploaded assets associated with that level. (Note that students will be blocked from deleting these assets in a future PR, so ignore the "delete" button.)
Screen Shot 2019-08-09 at 9 56 41 AM

### Next steps
- As a levelbuilder in levelbuilder_mode, I can:
  - Upload new starter assets in start mode
  - Delete starter assets in start mode

### CODE REVIEW

**Comment by davidsbailey** [comment]
👀 looking...

**Comment by davidsbailey** [comment]
Hi Maddie, here is some initial feedback. Sorry to get this to you so late on a Friday...

> we will request the starter assets associated with that levelChannelId from /level_starter_assets/:level_channel_id

this is a great pattern, but not a good name. "channel id" is something which corresponds to an entry in the storage_apps table in the pegasus DB, which is therefore environment-specific, so we shouldn't call this a channel id.

one idea for an alternative is to use the level name. This is appealing because the level name is consistent for a given level as it moves across environments. If you don't want to use the level name (one drawback is that renaming the level would lose the starter assets), then you could use a guid which gets stored in the level definition and you could call it something like "starter_assets_guid".

> levelbuilders will only be allowed to upload and delete starter assets on levelbuilder while in start mode

This sounds good!

**Comment by maddiedierker** [comment]
@davidsbailey you're definitely right about using the term "channel id" -- i thought it would be less confusing, but "starter_assets_guid" is definitely clearer and less overloaded. i'll update to use that terminology instead!

File: dashboard/app/controllers/level_starter_assets_controller.rb [hunk 1]
```diff
@@ -0,0 +1,69 @@
+class LevelStarterAssetsController < ApplicationController
+  S3_BUCKET = 'cdo-v3-assets'.freeze
+  S3_PREFIX = 'starter_assets/'.freeze
+
+  # GET /level_starter_assets/:id
+  def show
+    level_channel_id = params[:id]
```
Comment (davidsbailey) [inline dashboard/app/controllers/level_starter_assets_controller.rb]: within dashboard, I'd suggest `level_starter_assets_id` as the name for this kind of id. if that's too long, you could do `starter_assets_id`.

File: apps/src/clientApi.js [hunk 2]
```diff
@@ -442,9 +433,44 @@ class FilesApi extends CollectionsApi {
     );
   }
 }
+
+class StarterAssetsApi {
+  getStarterAssets(levelChannelId, onSuccess, onFailure) {
+    return ajaxInternal(
+      'GET',
+      this.withLevelChannelId(levelChannelId).basePath(''),
```
Comment (davidsbailey) [inline apps/src/clientApi.js]: within js code, `levelStarterAssetsId` could be a good alternative (to eliminate `ChannelId`).

**Review by davidsbailey — COMMENTED** [review]
started to add a few more comments on the same topic -- I think renaming "level channel id" to "level starter asset id" throughout will be a good improvement. I think this is true independently of whether you use the level name, a guid, or some other system for generating ids. 

I don't think it needs to block this PR, but we should nail down where these IDs will come from before too long.

**Comment by maddiedierker** [comment]
i have updated this PR with initial feedback, so PTAL!

File: dashboard/app/controllers/level_starter_assets_controller.rb [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  end
+
+  # GET /level_starter_assets/:level_name/:filename
+  # Returns requested file body as an IO stream.
+  def file
+    friendly_name = "#{params[:filename]}.#{params[:format]}"
+    guid_name = @level.starter_assets[friendly_name]
+    file_obj = get_file_object(prefix(guid_name))
+    content_type = file_content_type(File.extname(guid_name))
+    send_data read_file(file_obj), type: content_type, disposition: 'inline'
```
Comment (uponthesun) [inline dashboard/app/controllers/level_starter_assets_controller.rb:36]: Does this download the entire file into memory on our server first before streaming it to the client? Is there a way to start streaming the content immediately instead?

**Review by uponthesun — COMMENTED** [review]
Do we expect to ever have a high amount of traffic to a level with starter assets? If so, can we do some analysis around the scalability of the feature?

File: dashboard/app/controllers/level_starter_assets_controller.rb (hunk 3 above)
Comment (maddiedierker) [reply dashboard/app/controllers/level_starter_assets_controller.rb:36]: I think the answer to your first question is yes, but here's more detail since I'm not positive:

The `read_file` is a helper that maps to `file_obj.get.body.read` -- `body` is returned from the S3 SDK as an [IO](https://ruby-doc.org/core-2.5.1/IO.html), then I'm using Ruby's [`IO#read`](https://ruby-doc.org/core-2.5.1/IO.html#method-i-read) method to extract the file data and Rails' [`send_data`](https://apidock.com/rails/ActionController/Streaming/send_data) method to send the file to the client.

Levelbuilders currently use starter assets by uploading an image through our FilesApi, then the level renders that image using the [`/v3/assets/:channel_id/:filename`](https://github.com/code-dot-org/code-dot-org/blob/161ab002042bd31600d2045678f1147f910c067d/shared/middleware/files_api.rb#L137-L144) endpoint, so this new endpoint is meant to mimic that FilesApi endpoint.

Comment (uponthesun) [reply dashboard/app/controllers/level_starter_assets_controller.rb:36]: Okay, it looks like that `IO#read` is reading the entire contents into a String, so I think the answer is yes.

Instead of passing the result of `file_obj.get.body.read` to `send_data`, does it work if you pass `file_obj.get.body` itself? I think that would avoid that extra step.

Comment (maddiedierker) [reply dashboard/app/controllers/level_starter_assets_controller.rb:36]: No, returning `file_obj.get.body` doesn't work, but I did mess around with using [`read_nonblock`](https://ruby-doc.org/core-2.6.3/IO.html#method-i-read_nonblock) instead of `read` and it seems to work:

```
file_obj.get.body.read_nonblock(file_obj.size)
```

but I don't feel like I really understand what read_nonblock does

File: apps/src/code-studio/components/AssetManager.jsx [hunk 4]
```diff
@@ -52,26 +56,48 @@ export default class AssetManager extends React.Component {
...
     this.state = {
       assets: null,
+      starterAssets: null,
       statusMessage: props.uploadsEnabled ? '' : errorUploadDisabled,
       recordingAudio: false,
       audioErrorType: AudioErrorType.NONE
     };
   }
 
   componentWillMount() {
```
Comment (islemaster) [inline apps/src/code-studio/components/AssetManager.jsx:78]: `componentWillMount` [is deprecated](https://reactjs.org/docs/react-component.html#mounting).  Maybe we should change this up while we're in here?

**Review by islemaster — APPROVED** [review]
This is awesome!

It does introduce one more feature that doesn't work for a local developer that isn't set up for S3 access, which we've run into with contributors a few times.  I wonder if there's a lightweight way to add a fallback while this feature is still new.

File: apps/src/code-studio/components/AssetManager.jsx (hunk 4 above)
Comment (maddiedierker) [reply apps/src/code-studio/components/AssetManager.jsx:78]: good call! i'll make a follow-up PR for this

File: dashboard/app/controllers/level_starter_assets_controller.rb (hunk 3 above)
Comment (davidsbailey) [reply dashboard/app/controllers/level_starter_assets_controller.rb:36]: Regarding performance, we should be able to set a very long cache lifetime when starter assets are requested for read, which should get us most of the way to good performance.

**Comment by davidsbailey** [comment]
> It does introduce one more feature that doesn't work for a local developer that isn't set up for S3 access, which we've run into with contributors a few times. I wonder if there's a lightweight way to add a fallback while this feature is still new.

Good call Brad, now seems like a good time to think through this. This are my initial thoughts on our current state of the world + current problems + potential solutions:

1. Today, when a volunteer needs to be able to save their work on project-backed levels locally, we give them access to the `cdo-v3-sources-contributors` bucket, which anyone can read or write to.

2. If a volunteer needed to access assets in applab as a normal user (not levelbuilder), we could give them access to a new `cdo-v3-assets-contributors` bucket with similar permissions.

3. If we do (2), but the volunteer also needs to run levels with starter assets, they wouldn't be able to, because those starter assets would not be present in `cdo-v3-assets-contributors`. possible solutions:

    A. whenever we upload a starter asset to `cdo-v3-assets`, also upload it to `cdo-v3-assets-contributors`, or

    B. in development,  when loading a level that has a starter asset, use a production url to the starter asset instead of a local one, unless a certain flag like `CDO.use_local_starter_asset_urls` is set in locals.yml . when that flag is not set, you don't need S3 keys to run levels with starter assets. when that flag is set, you do need S3 keys. Also, you shouldn't be able to upload (or delete?) starter assets locally unless `CDO.use_local_starter_asset_urls` is set.

related problem:
* if I run my local machine in levelbuilder_mode, it could be useful to have some safeguards so that I am not uploading files to the same directory that is used by production, so that if there is a bug in my code, I don't accidentally delete or overwrite any existing starter assets. I think the same `CDO.use_local_starter_asset_urls` flag from option 3B could be used to solve this.

I think option 2 + 3B might give us all the things we care about, but @islemaster and @madelynkasula , please poke some holes in my reasoning or let me know if it seems over-engineered.

File: apps/src/clientApi.js [hunk 5]
```diff
@@ -442,9 +433,44 @@ class FilesApi extends CollectionsApi {
     );
   }
 }
+
+class StarterAssetsApi {
+  getStarterAssets(levelName, onSuccess, onFailure) {
+    return ajaxInternal(
+      'GET',
+      this.withLevelName(levelName).basePath(''),
```
Comment (davidsbailey) [inline apps/src/clientApi.js:441]: this looks like a great pattern!

**Comment by maddiedierker** [comment]
those are all great thoughts for possible solutions. first, i want to talk through this part:

> if I run my local machine in levelbuilder_mode, it could be useful to have some safeguards so that I am not uploading files to the same directory that is used by production, so that if there is a bug in my code, I don't accidentally delete or overwrite any existing starter assets.

currently, the safeguards would be:
- uploading a new file generates a UUID filename so that starter assets will not be overridden in S3 due to name collisions (so all the files in S3 are formatted as `cdo-v3-assets/starter_assets/123-abc.png`)
- [currently in-progress] deleting an asset will remove that asset from the .level file definition, but not from S3. 

TLDR; since the .level file stores the current state of starter assets, devs can upload/delete assets locally without affecting production, _as long as they do not commit the changes to the .level file_. however, this does mean that all starter assets are stored in the bucket used by production.

so my question is, are those safeguards enough? or should we go down the path of storing assets for different environments in different buckets?

**Comment by davidsbailey** [comment]
Hey @madelynkasula , good points, I think you've convinced me that the risk of this feature messing with production data when running on localhost is relatively low. let's not worry about separating starter assets between localhost and production (+ other environments) unless it is necessary to do so in order to solve the volunteer scenario.

File: apps/src/clientApi.js [hunk 6]
```diff
@@ -442,9 +433,44 @@ class FilesApi extends CollectionsApi {
...
+      const error =
+        'You must bind the API and set levelName before creating a base path.';
+      throw new Error(error);
+    }
+
+    return `/level_starter_assets/${this.levelName}/${path}`;
+  }
+
+  deleteFile(filename, success, error) {
+    throw new Error('not yet implemented');
```
Comment (davidsbailey) [inline apps/src/clientApi.js:464]: 👍 good safeguard!

File: apps/src/code-studio/components/AssetRow.jsx [hunk 7]
```diff
@@ -41,6 +45,15 @@ export default class AssetRow extends React.Component {
     attemptedUsedDelete: false
   };
 
+  api = () => {
+    const {api, useFilesApi} = this.props;
+    if (api) {
+      return api;
+    } else {
+      return useFilesApi ? filesApi : assetsApi;
```
Comment (davidsbailey) [inline apps/src/code-studio/components/AssetRow.jsx:53]: nit: it's a bit weird to have the api-picking logic split like this. I would recommend moving it entirely outside of (or entirely inside of) The AssetRow. fine to consider doing in a later PR

File: dashboard/app/controllers/level_starter_assets_controller.rb (hunk 3 above)
Comment (maddiedierker) [reply dashboard/app/controllers/level_starter_assets_controller.rb:36]: thanks, dave! winter suggested the same, so https://github.com/code-dot-org/code-dot-org/pull/30218/commits/16b3f4968ec77d28267d264df9cdc8dd05d34502 adds a 1-hour cache expiration to the response using [expires_in](https://apidock.com/rails/ActionController/ConditionalGet/expires_in), which sets the Cache-Control header. 

is this enough to cache the response in Cloudfront? is there a better/standardized way to do this?

also, i grabbed some response headers if folks are interested--

response headers from initial request to `/level_starter_assets/:level_name/:filename` with Cache-Control being set:
```
HTTP/1.1 200 OK
Content-Type: image/png
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Date: Mon, 19 Aug 2019 21:15:32 GMT
Content-Disposition: inline
Content-Transfer-Encoding: binary
Cache-Control: max-age=3600, public
ETag: W/"cc9dc8917ec61f2a02815a6cb26f4fae"
Content-Length: 14699
Vary: Host
X-Content-Digest: 537722607b2294661f42b967a1bb64e2a4299822
Age: 1
X-Rack-Cache: miss, store
X-Request-Id: 59b4b502-410e-4d16-9157-499cc7de00a1
X-Runtime: 0.454960
Connection: keep-alive
Server: thin
```

response headers from a subsequent request:
```
HTTP/1.1 200 OK
Content-Type: image/png
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Date: Mon, 19 Aug 2019 21:15:32 GMT
Content-Disposition: inline
Content-Transfer-Encoding: binary
Cache-Control: max-age=3600, public
ETag: W/"cc9dc8917ec61f2a02815a6cb26f4fae"
Content-Length: 14699
Vary: Host
X-Content-Digest: 537722607b2294661f42b967a1bb64e2a4299822
Age: 105
X-Rack-Cache: fresh
X-Request-Id: f7d97d50-fc79-4b61-b1a0-733961f484d0
X-Runtime: 0.124507
Connection: keep-alive
Server: thin
```

Comment (davidsbailey) [reply dashboard/app/controllers/level_starter_assets_controller.rb:36]: I'm not 100% sure this applies here, but just wanted to share a pointer to a different place where we have to read the entire file rather than streaming it to the client: https://github.com/code-dot-org/code-dot-org/blob/a26a95f71fc20f83fb0b24bd4b9629d550bb9016/shared/middleware/files_api.rb#L405-L409

this might not be relevant here yet since we're not conditionally rejecting the request, also this is rails and not just sinatra etc. but wanted to include the link in case the problem described there comes up here.

**Review by davidsbailey — APPROVED** [review]
This looks great, Maddie! Nice work!

File: apps/src/code-studio/components/AssetRow.jsx (hunk 7 above)
Comment (maddiedierker) [reply apps/src/code-studio/components/AssetRow.jsx:53]: good call! i'll address in a follow-up PR

File: dashboard/app/controllers/level_starter_assets_controller.rb (hunk 3 above)
Comment (davidsbailey) [reply dashboard/app/controllers/level_starter_assets_controller.rb:36]: `expires_in` looks right, and the cache headers look good to me.

there may be one more step to make this route cacheable. you need to tell varnish / cloudfront to ignore cookies and headers, so that it doesn't try to cache a separate copy for every user, language, etc. here is an example of how to do that: https://github.com/code-dot-org/code-dot-org/blob/d16fc9addb88600299d92889c29a019e722f5089/cookbooks/cdo-varnish/libraries/http_cache.rb#L222-L229

I'm not really sure how it works when you want the same route to be non-cacheable for PUT or POST requests but be cacheable for GET requests. that's why we used a different route like `/v3/files-public` instead of `/v3/files` at the above link. so one idea would be to split the API so that the publicly cacheable route has a different prefix.

TBH I think this caching change is getting big enough that it might be better to split it out into another PR. it's fine to leave `expires_in` how you have it but just keep in mind that you might have more work to do before it takes effect.

Comment (maddiedierker) [reply dashboard/app/controllers/level_starter_assets_controller.rb:36]: ohh, interesting -- thank you! yeah, that sounds like it should belong in its own PR. winter and i also discussed refactoring the client so that it has a reference to the assets' S3 URLs so that we can hit S3 directly, rather than having separate endpoints on our servers that just serve content from S3. so i think i will size that work up and probably go that route, since it will have the same benefit and will also remove complexity from this feature
